clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3786,15 +3786,17 @@ public ListOffsetsResult listOffsets(Map<TopicPartition, OffsetSpec> topicPartit
                 void handleResponse(AbstractResponse abstractResponse) {
                     ListOffsetResponse response = (ListOffsetResponse) abstractResponse;
                     Set<TopicPartition> partitionsWithErrors = new HashSet<>();
+                    Map<TopicPartition, OffsetSpec> retryTopicPartitionOffsets = new HashMap<>();
 
                     for (Entry<TopicPartition, PartitionData> result : response.responseData().entrySet()) {
                         TopicPartition tp = result.getKey();
                         PartitionData partitionData = result.getValue();
 
                         KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(tp);
                         Errors error = partitionData.error;
-                        if (MetadataOperationContext.shouldRefreshMetadata(error)) {
+                        if (MetadataOperationContext.shouldRefreshMetadata(error) && topicPartitionOffsets.get(tp) != null) {
```

Comment:
This code reads a bit awkwardly because we only have the check for the first branch. It seems like we are trying to handle the case that the broker has sent us responses for partitions that we did not ask for. Perhaps that should be a fatal error and we can call `completeExceptionally`?

Reply:
Yeah I was trying to write the code so that we ignore any partitions we haven't asked for. Is it better behavior in this case to just silently ignore or raise an error and `completeExceptionally`?